File: 单调栈/4. 接雨水 - 42/Trapping Rain Water.cpp
/*
    23/03/27
    https://leetcode.cn/problems/trapping-rain-water/
 */

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

/*
    solution: 栈
    result: ✅
    TC: O(n)
    SC: O(n)
 */
int trap3(vector<int>& height) {

    int lengthOfHeight = height.size();
    int sumRain = 0;
    stack<int> st;

    for (int i = 0; i < lengthOfHeight; i ++) {
        while (!st.empty() && height[i] > height[st.top()]) {
            int mid_height = st.top();
            st.pop();
            if (!st.empty()) {
                int h = min(height[st.top()], height[i]) - height[mid_height];
                int w = i - st.top() - 1;
                sumRain += h * w;
            }
        }
        st.push(i);
    }

    return sumRain;
}

/*
    solution: 双指针
    result: ✅
    TC: O(n)
    SC: O(n)
 */
int trap2(vector<int>& height) {

    int lengthOfHeight = height.size();
    int sumRain = 0;
    vector<int> leftMaxHeight(height.size(), 0);
    vector<int> rightMaxHeight(height.size(), 0);
    leftMaxHeight[0] = height[0];
    rightMaxHeight[lengthOfHeight-1] = height[lengthOfHeight-1];

    // store max height of all elements from the left 
    for (int i = 1; i < lengthOfHeight; i ++) {
        leftMaxHeight[i] = max(height[i], leftMaxHeight[i-1]);
    }
    // store max height of all elements from the left 
    for (int i = lengthOfHeight - 2; i >= 0; i --) {
        rightMaxHeight[i] = max(height[i], rightMaxHeight[i+1]);
    }

    for (int i = 0; i < lengthOfHeight; i ++) {
        if (i == 0 || i == lengthOfHeight - 1) { continue; }

        int tmpSum = min(leftMaxHeight[i], rightMaxHeight[i]) - height[i];
        sumRain = tmpSum > 0 ? sumRain + tmpSum : sumRain;
    }

    return sumRain;

}


/*
    solution: 暴力
    result: ❌ 用例均通过，但超时 
    TC: O(n^2)
    SC: O(1)
 */
int trap1(vector<int>& height) {

    int lengthOfHeight = height.size();
    int sumRain = 0;
    
    for (int i = 0; i < lengthOfHeight; i++)
    {
        if (i == 0 || i == lengthOfHeight - 1){ continue; } 

        auto lh = max_element(height.begin(), height.begin() + i);
        auto rh = max_element(height.begin() + i + 1, height.end());
        int tmpSum = min(*lh, *rh) - height[i];
        sumRain = tmpSum > 0 ? sumRain + tmpSum : sumRain;
    }

    return sumRain;

}

File: 单调栈/4. 接雨水 - 42/Trapping Rain Water.py
# 23/03/27
# https://leetcode.cn/problems/trapping-rain-water/

from collections import List

# 栈
# ✅

def trap3(height: List[int]) -> int:

    lengthOfHeight = len(height)
    sumRain = 0
    stack = list()

    for i in range(lengthOfHeight):
        while stack and height[i] > height[stack[-1]]:
            mid_height = stack.pop()
            if stack:
                h = min(height[stack[-1]], height[i]) - height[mid_height]
                w = i - stack[-1] - 1
                sumRain += w * h 
        stack.append(i)
    return sumRain

# 双指针
# ✅

def trap2(height: List[int]) -> int: 

    lengthOfHeight = len(height)
    sumRain = 0
    leftMaxHeight, rightMaxHeight = [0] * lengthOfHeight, [0] * lengthOfHeight
    leftMaxHeight[0], rightMaxHeight[lengthOfHeight - 1] = height[0], height[lengthOfHeight -1]

    for i in range(1, lengthOfHeight):
        leftMaxHeight[i] = max(height[i], leftMaxHeight[i-1])
    for i in range(lengthOfHeight - 2, -1, -1):
        rightMaxHeight[i] = max(height[i], rightMaxHeight[i+1])
    for i in range(lengthOfHeight):
        if i == 0 or i == lengthOfHeight - 1: continue
        tmpSum = min(leftMaxHeight[i], rightMaxHeight[i]) - height[i]
        sumRain = tmpSum +  sumRain if tmpSum > 0 else sumRain
    
    return sumRain

    

# 暴力
# ❌ 超时，用例：https://leetcode.cn/submissions/detail/418149515/testcase/

def trap1(height: List[int]) -> int: 

    lengghtOfHeight = len(height)
    sumRain = 0

    for i in range(lengghtOfHeight):

        if i == 0 or i == lengghtOfHeight - 1: continue

        lh, rh = max(height[:i]), max(height[(i+1):])
        tmpSum = min(lh, rh) - height[i]
        sumRain = sumRain + tmpSum if tmpSum > 0 else sumRain
    
    return sumRain
        